File: src/Integration/PaymentService.php
<?php
declare(strict_types=1);
namespace App\Integration;
use App\Billing\{BillingError,BillingService};
use App\Infrastructure\{Database,JobDeferred};
use App\Infrastructure\WebhookGuard;
use App\Integration\Payment\ProviderRegistry;
use App\Payments\PaymentEventStore;
use App\Payments\PaymentAttemptStore;
use App\Observability\OperationsService;
use Symfony\Contracts\HttpClient\HttpClientInterface;
/**
 * Unified payment service. Routes order/topup checkouts and verification
 * through the provider registry. Keeps the legacy Payments API for
 * compatibility while delegating to providers.
 */
final class PaymentService
{
    public function __construct(
        private Database $db,
        private BillingService $billing,
        private HttpClientInterface $http,
        private array $config,
        private ProviderRegistry $registry,
        private ?PaymentEventStore $events=null,
        private ?WebhookGuard $webhookGuard=null,
        private ?\App\Integration\Mailer $mailer=null,
        private ?PaymentAttemptStore $attempts=null,
    ) {}
    public function registry(): ProviderRegistry { return $this->registry; }
    /** Create checkout for an order. Returns [payment_id, checkout_url]. */
    public function createOrder(string $orderId): array
    {
        $order = $this->db->one('SELECT * FROM orders WHERE id=?', [$orderId]);
        if (!$order || $order['status'] !== 'pending' || $order['checkout_url']) return ['payment_id' => $order['provider_payment_id'] ?? '', 'checkout_url' => $order['checkout_url'] ?? ''];
        $attempt=$this->attempts?->begin('order',$order);
        if ($attempt && !($attempt['created']??false)) {
            if (!empty($attempt['provider_payment_id']) && !empty($attempt['checkout_url'])) {
                // A prior worker owns this durable attempt and completed the
                // provider call. Restore the legacy projection if needed.
                $this->db->execute('UPDATE orders SET provider_payment_id=?,checkout_url=? WHERE id=? AND checkout_url IS NULL',[$attempt['provider_payment_id'],$attempt['checkout_url'],$orderId]);
                return ['payment_id'=>$attempt['provider_payment_id'],'checkout_url'=>$attempt['checkout_url']];
            }
            // Do not issue a second mutating provider request while its result
            // is unknown. Recovery/verification must resolve the first intent.
            throw new JobDeferred(60);
        }
        if ($order['provider'] === 'demo') {
            if (($this->config['APP_ENV'] ?? 'dev') === 'prod') throw new BillingError('Демоплатёж запрещён.');
            $this->db->execute('UPDATE orders SET provider_payment_id=?,checkout_url=? WHERE id=?',['demo_'.$orderId,'/orders/'.$orderId,$orderId]);
            if($attempt)$this->attempts->attached($attempt['id'],'demo_'.$orderId,'/orders/'.$orderId);
            return ['payment_id' => 'demo_'.$orderId, 'checkout_url' => '/orders/'.$orderId];
        }
        $this->requireCheckoutSupport($order['provider']);
        $provider = $this->registry->get($order['provider']);
        $user = $this->db->one('SELECT * FROM users WHERE id=?', [$order['user_id']]);
        try {$result = $provider->createOrder($order, $user ?? []);} catch (\Throwable $e) { if($attempt)$this->attempts->unknown($attempt['id'],$e); throw $e; }
        $this->db->execute('UPDATE orders SET provider_payment_id=?,checkout_url=? WHERE id=? AND (provider_payment_id IS NULL OR provider_payment_id=?)', [$result['payment_id'], $result['checkout_url'], $orderId, $result['payment_id']]);
        if($attempt)$this->attempts->attached($attempt['id'],$result['payment_id'],$result['checkout_url']);
        return $result;
    }
    /** Create checkout for a topup. Returns [payment_id, checkout_url]. */
    public function createTopup(string $topupId): array
    {
        $topup = $this->db->one('SELECT * FROM topups WHERE id=?', [$topupId]);
        if (!$topup || $topup['status'] !== 'pending' || $topup['checkout_url']) return ['payment_id' => $topup['provider_payment_id'] ?? '', 'checkout_url' => $topup['checkout_url'] ?? ''];
        $attempt=$this->attempts?->begin('topup',$topup);
        if ($attempt && !($attempt['created']??false)) {
            if (!empty($attempt['provider_payment_id']) && !empty($attempt['checkout_url'])) {
                $this->db->execute('UPDATE topups SET provider_payment_id=?,checkout_url=? WHERE id=? AND checkout_url IS NULL',[$attempt['provider_payment_id'],$attempt['checkout_url'],$topupId]);
                return ['payment_id'=>$attempt['provider_payment_id'],'checkout_url'=>$attempt['checkout_url']];
            }
            throw new JobDeferred(60);
        }
        if ($topup['provider'] === 'demo') {
            if (($this->config['APP_ENV'] ?? 'dev') === 'prod') throw new BillingError('Демоплатёж запрещён.');
            $this->db->execute('UPDATE topups SET provider_payment_id=?,checkout_url=? WHERE id=?',['demo_'.$topupId,'/balance/topup/'.$topupId,$topupId]);
            if($attempt)$this->attempts->attached($attempt['id'],'demo_'.$topupId,'/balance/topup/'.$topupId);
            return ['payment_id' => 'demo_'.$topupId, 'checkout_url' => '/balance/topup/'.$topupId];
        }
        $this->requireCheckoutSupport($topup['provider']);
        $provider = $this->registry->get($topup['provider']);
        $user = $this->db->one('SELECT * FROM users WHERE id=?', [$topup['user_id']]);
        try {$result = $provider->createTopup($topup, $user ?? []);} catch (\Throwable $e) { if($attempt)$this->attempts->unknown($attempt['id'],$e); throw $e; }
        $this->db->execute('UPDATE topups SET provider_payment_id=?,checkout_url=? WHERE id=? AND (provider_payment_id IS NULL OR provider_payment_id=?)', [$result['payment_id'], $result['checkout_url'], $topupId, $result['payment_id']]);
        if($attempt)$this->attempts->attached($attempt['id'],$result['payment_id'],$result['checkout_url']);
        return $result;
    }
    private function requireCheckoutSupport(string $provider): void
    {
        // Only Platega is supported after the provider cleanup.
    }
    /** Verify a payment with the provider and settle if paid. */
    public function verify(string $paymentId, ?string $providerId=null, ?string $correlationId=null): void
    {
        if ($paymentId==='' || strlen($paymentId)>100) throw new BillingError('Некорректный платёж.');
        $params=[$paymentId,$paymentId];
        // Keep the provider restriction around *both* lookup alternatives.
        // Without parentheses SQL applies AND only to `id=?`, allowing a
        // colliding provider_payment_id from another provider to be selected.
        $scope=$providerId!==null?' AND provider=?':'';
        if ($providerId!==null) $params[]=$providerId;
        $orders=$this->db->all("SELECT * FROM orders WHERE (provider_payment_id=? OR id=?)".$scope,$params);
        $topups=$this->db->all("SELECT * FROM topups WHERE (provider_payment_id=? OR id=?)".$scope,$params);
        if (count($orders)+count($topups)>1) throw new BillingError('Неоднозначная привязка платежа.');
        $entity=$orders[0]??$topups[0]??null;
        $providerId??=$entity['provider']??null;
        if ($providerId===null || $providerId==='demo') return;
        $result=$this->registry->get($providerId)->verify($paymentId);
        if (!in_array($result['status']??'',['paid','canceled'],true)) {
            if ($correlationId!==null) throw new JobDeferred(60);
            return;
        }
        $metadata=$result['metadata']??[];
        // Recover a checkout that succeeded remotely before its id was stored locally.
        if (!$entity && $providerId==='platega') {
            $orderId=$metadata['order_id']??$metadata['merchant_order_id']??'';
            $topupId=$metadata['topup_id']??$metadata['merchant_order_id']??'';
            $order=$this->db->one('SELECT * FROM orders WHERE id=? AND provider=?',[$orderId,$providerId]);
            $topup=$this->db->one('SELECT * FROM topups WHERE id=? AND provider=?',[$topupId,$providerId]);
            if ($order && $topup) throw new BillingError('Неоднозначная привязка платежа.');
            $entity=$order??$topup;
            $orders=$order?[$order]:[];
        }
        if (!$entity) throw new BillingError('Платёж пока не привязан к локальному заказу. Требуется сверка.');
        $isOrder=$orders!==[];
        // When the entity was resolved by its own stored provider_payment_id,
        // the binding is already authoritative (payment belongs to this entity).
        // Metadata-id checks only matter for the recover-checkout path (entity
        // matched via metadata, not by provider_payment_id). A canceled/expired
        // Platega tx returns no orderId, so an empty metadata id must not reject
        // a topup/order whose provider_payment_id already equals the payment.
        $boundById = ($entity['provider_payment_id'] ?? null) === (string)$paymentId;
        $expected = $metadata[$isOrder ? 'order_id' : 'topup_id'] ?? null;
        if (!$boundById
            && $providerId==='platega'
            && $expected !== $entity['id']) throw new BillingError('Платёж относится к другому заказу.');
        $account=match($providerId){'platega'=>($this->config['PLATEGA_MERCHANT_ID']??''),default=>''};
        if (isset($entity['provider_account']) && $entity['provider_account']!=='' && $entity['provider_account']!==$account) throw new BillingError('Несовпадение магазина.');
        $actualId=$result['payment_id']??'';
        if (!is_string($actualId) || $actualId==='' || strlen($actualId)>100 || ($entity['provider_payment_id']!==null && $entity['provider_payment_id']!==$actualId)) throw new BillingError('Несовпадение платежа.');
        if (($result['status']??'')==='paid') {
            if (($this->config['APP_ENV']??'dev')==='prod' && !empty($result['test'])) throw new BillingError('Тестовый платёж запрещён в production.');
            if ($isOrder) $this->billing->settle($entity['id'],$providerId,$actualId,(int)$result['amount_kopeks'],$result['currency'],$correlationId);
            else $this->billing->settleTopup($entity['id'],$providerId,$actualId,(int)$result['amount_kopeks'],$result['currency']);
            $this->attempts?->completed($providerId,$actualId,'paid');
        } elseif (($result['status']??'')==='canceled') {
            $table=$isOrder?'orders':'topups';
            $this->db->execute("UPDATE ".$table." SET status='canceled' WHERE id=? AND provider=? AND status='pending'",[$entity['id'],$providerId]);
            $this->attempts?->completed($providerId,$actualId,'canceled');
        }
    }
    /** Handle a provider webhook. Returns true if the webhook was ours. */
    public function handleWebhook(string $providerId, \Symfony\Component\HttpFoundation\Request $request): bool
    {
        if (!$this->registry->has($providerId)) return false;
        $provider = $this->registry->get($providerId);
        // A webhook must never wake a retry loop for a provider whose own
        // verification credentials are absent. Bad JSON is an invalid
        // delivery, not an application error worthy of a 500 response.
        if (!$provider->configured()) return false;
        try {
            $result = $provider->handleWebhook($request);
        } catch (\Throwable) {
            return false;
        }
        if ($result === null) return false;
        $paymentId = $result['payment_id'];
        $status = $result['status'];
        if (!is_string($paymentId) || $paymentId==='' || strlen($paymentId)>100) return false;
        if (in_array($status,['paid','canceled'],true)) {
            // Acknowledge only after a durable inbox row and its outbox command
            // are committed. The provider API remains authoritative for settlement.
            $this->db->transaction(function () use ($paymentId,$providerId,$status,$result,$request) {
                $eventId=(string)($result['event_id'] ?? ($paymentId.':'.$status));
                // Tamper guard (only defense in depth since Platega webhooks
                // carry no HMAC/timestamp): same provider_event_id but changed
                // payload content is a tamper attempt. WebhookGuard::claim()
                // returns 'new' | 'duplicate'(same payload, already processed)
                // | 'same_payload'(not yet processed) and THROWS on an id reuse
                // with different content — that exception is caught below and
                // surfaced without touching money.
                if ($this->webhookGuard !== null) {
                    try {
                        $claim = $this->webhookGuard->claim($providerId, $eventId, $result);
                    } catch (\RuntimeException $e) {
                        // Tamper: same event id, different payload. Log and raise
                        // an Ops alarm (read-only) but acknowledge the webhook so
                        // an attacker can't force a 500-loop; money is untouched.
                        error_log('webhook.tamper '.$providerId.'/'.$eventId.' '.$e->getMessage());
                        try {
                            $ops=(new OperationsService($this->db));
                            $corr='webhook-tamper:'.$providerId.':'.$eventId;
                            $op=$ops->start('webhook.tamper',['metadata'=>['provider'=>$providerId,'event_id'=>$eventId]],$corr);
                            if(!$op['existing']){
                                $ops->event($op['id'],'webhook.tamper.detected','failed','Webhook payload mismatch (same id, different content) for '.$providerId.'/'.substr($eventId,0,40).' — possible tamper.',['metadata'=>['provider'=>$providerId,'event_id'=>$eventId]]);
                                $ops->complete($op['id'],'failed','Проверить руками');
                                // Deliver a human-visible alert through the working
                                // SMTP pipeline (queue -> Reconciler flushQueue -> send).
                                if ($this->mailer !== null) {
                                    $adminEmail=$this->db->one("SELECT email FROM users WHERE role='admin' AND email IS NOT NULL ORDER BY created_at LIMIT 1")['email']??'';
                                    if ($adminEmail !== '') {
                                        try {
                                            $this->mailer->queue($adminEmail,
                                                '[ASTRACAT] Webhook tamper: '.$providerId,
                                                'Webhook payload mismatch (same id, different content) for '.$providerId.'/'.substr($eventId,0,40).'<br>Событие помечено processed=2 (tamper). Проверьте, не идёт ли атака на вебхуки.'
                                            );
                                        } catch (\Throwable) {}
                                    }
                                }
                            }
                        } catch (\Throwable) {}
                        return;
                    }
                    if ($claim === 'duplicate') {
                        return; // same payload, already processed => benign replay
                    }
                }
                $id=$this->events?->receive($providerId,$eventId,$paymentId,$result,true,$this->webhookHeaders($request)) ?? '';
                if ($id!=='') {
                    // Replay guard: if this event was already fully processed, a
                    // repeated delivery of the same provId/eventId is a benign
                    // duplicate (or a replay) — ack it but do NOT re-enqueue a
                    // second outbox command, or settlement state could be
                    // re-evaluated on the same event. Only first-seen (not yet
                    // processed) events spawn a processing job.
                    $already = $this->db->one(
                        "SELECT status FROM payment_events WHERE id=?", [$id]
                    );
                    if ($already !== null && $already['status'] === 'processed') {
                        return; // benign replay/duplicate: already handled
                    }
                    $this->db->execute('INSERT INTO outbox(id,topic,dedup_key,payload,priority,available_at,created_at,correlation_id) VALUES(?,?,?,?,?,?,?,?) ON CONFLICT(dedup_key) DO NOTHING',[
                        Database::id(),'payment.event.process','payment-event:'.$id,json_encode(['event_id'=>$id],JSON_THROW_ON_ERROR),100,time(),time(),$id
                    ]);
                } else $this->outboxEnqueueVerify($paymentId,$providerId,$status);
            });
        }
        return true;
    }
    /** Worker-only: process one durable webhook event exactly once locally. */
    public function processEvent(string $eventId): void
    {
        $event=$this->events
            ? $this->events->claim($eventId)
            : $this->db->one('SELECT * FROM payment_events WHERE id=? AND processed_at IS NULL AND signature_valid=1',[$eventId]);
        if (!$event) return;
        $operations=new OperationsService($this->db);
        $correlation='cor_'.substr(hash('sha256',$event['provider'].':'.($event['payment_id']?:$event['provider_event_id'])),0,40);
        $op=$this->db->one('SELECT id FROM operations WHERE correlation_id=?',[$correlation]);
        try {
            // Webhooks are hints, including cancellations and events arriving
            // before checkout persistence. Always consult the provider before
            // changing state; API metadata can recover a lost local binding.
            $pid=(string)$event['payment_id'];
            $this->verify($pid,$event['provider'],$correlation);
            if ($this->events) $this->events->processed($eventId,(string)$event['lock_token']);
            if($op){$operations->event($op['id'],'payment.verified','success','Payment verified with provider');$operations->complete($op['id']);}
        } catch (\Throwable $e) {
            if ($this->events) $this->events->failed($eventId,(string)$event['lock_token'],$e);
            if($op)$operations->fail($op['id'],$e);
            throw $e;
        }
    }
    /** Reconciliation recovery for inbox events whose outbox delivery died or lease expired. */
    public function requeueStaleEvents(int $limit=100): int
    {
        // Unknown checkout creation is deliberately never replayed: the
        // provider might have accepted the original request. Surface stale
        // outcomes for manual reconciliation instead of silently deferring
        // them forever.
        $this->attempts?->escalateUnknown(900,$limit);
        $now=time();
        $rows=$this->db->all("SELECT id FROM payment_events WHERE signature_valid=1 AND ((status IN ('pending','retry') AND COALESCE(next_attempt_at,received_at)<=?) OR (status='processing' AND locked_until<=?)) ORDER BY received_at LIMIT ?",[$now,$now,$limit]);
        $bucket=intdiv($now,300);
        foreach($rows as $row) $this->db->execute('INSERT INTO outbox(id,topic,dedup_key,payload,priority,available_at,created_at,correlation_id) VALUES(?,?,?,?,?,?,?,?) ON CONFLICT(dedup_key) DO NOTHING',[Database::id(),'payment.event.process','payment-event-recover:'.$row['id'].':'.$bucket,json_encode(['event_id'=>$row['id']],JSON_THROW_ON_ERROR),100,$now,$now,$row['id']]);
        return count($rows);
    }
    private function outboxEnqueueVerify(string $paymentId,string $providerId,string $status): void
    {
        // Reuse the outbox for authoritative verification (webhook is only a hint).
        // payment.verify is the highest-priority job; hardcode 100 to match Outbox::PRIORITY.
        $this->db->execute('INSERT INTO outbox(id,topic,dedup_key,payload,priority,available_at,created_at) VALUES(?,?,?,?,?,?,?) ON CONFLICT(dedup_key) DO NOTHING', [
            Database::id(), 'payment.verify', 'verify:'.$providerId.':'.$paymentId.':'.$status.':'.intdiv(time(),60), json_encode(['payment_id' => $paymentId,'provider'=>$providerId], JSON_THROW_ON_ERROR), 100, time(), time(),
        ]);
    }

    /** Evidence only: never retain Authorization, cookies, or provider secrets. */
    private function webhookHeaders(\Symfony\Component\HttpFoundation\Request $request): array
    {
        $allowed=['content-type','x-request-id','x-signature','x-signature-sha256','x-webhook-id'];
        $headers=[];
        foreach($allowed as $name) if($request->headers->has($name)) $headers[$name]=$request->headers->all($name);
        return $headers;
    }
}

File: rails_backend/app/services/payments/payment_service.rb
module Payments
  class PaymentService
    def initialize(
      settlement: Billing::SettlementService.new,
      topups: Billing::TopupService.new,
      attempts: PaymentAttemptStore.new,
      events: PaymentEventStore.new,
      outbox: Infrastructure::OutboxService.new,
      webhook_guard: WebhookGuard.new
    )
      @settlement = settlement
      @topups = topups
      @attempts = attempts
      @events = events
      @outbox = outbox
      @webhook_guard = webhook_guard
    end

    def create_order(order_id)
      order = Order.find_by(id: order_id)
      return { "payment_id" => "", "checkout_url" => "" } unless order
      Infrastructure::SafetyControls.new.assert_can_purchase!(order.provider)
      return { "payment_id" => order.provider_payment_id.to_s, "checkout_url" => order.checkout_url.to_s } unless order.pending? && order.checkout_url.blank?

      attempt = @attempts.begin_attempt("order", order)
      return existing_attempt_result(order, attempt) unless attempt.created?

      if order.provider == "demo"
        raise Billing::BillingError, "Демоплатёж запрещён." if production?
        return attach_order_checkout(order, attempt, "demo_#{order.id}", "/orders/#{order.id}")
      end

      provider = provider_for(order.provider)
      result = provider.create_order(order, order.user)
      attach_order_checkout(order, attempt, result["payment_id"], result["checkout_url"])
    rescue StandardError => e
      @attempts.unknown(attempt.id, e) if defined?(attempt) && attempt
      raise
    end

    def create_topup(topup_id)
      topup = Topup.find_by(id: topup_id)
      return { "payment_id" => "", "checkout_url" => "" } unless topup
      Infrastructure::SafetyControls.new.assert_can_purchase!(topup.provider)
      return { "payment_id" => topup.provider_payment_id.to_s, "checkout_url" => topup.checkout_url.to_s } unless topup.pending? && topup.checkout_url.blank?

      attempt = @attempts.begin_attempt("topup", topup)
      return existing_attempt_result(topup, attempt) unless attempt.created?

      if topup.provider == "demo"
        raise Billing::BillingError, "Демоплатёж запрещён." if production?
        return attach_topup_checkout(topup, attempt, "demo_#{topup.id}", "/balance/topup/#{topup.id}")
      end

      provider = provider_for(topup.provider)
      result = provider.create_topup(topup, topup.user)
      attach_topup_checkout(topup, attempt, result["payment_id"], result["checkout_url"])
    rescue StandardError => e
      @attempts.unknown(attempt.id, e) if defined?(attempt) && attempt
      raise
    end

    def verify(payment_id, provider_id = nil, correlation_id: nil)
      raise Billing::BillingError, "Некорректный платёж." if payment_id.blank? || payment_id.length > 100

      orders = scoped_payment_lookup(Order, payment_id, provider_id)
      topups = scoped_payment_lookup(Topup, payment_id, provider_id)
      raise Billing::BillingError, "Неоднозначная привязка платежа." if orders.size + topups.size > 1

      entity = orders.first || topups.first
      provider_id ||= entity&.provider
      return if provider_id.blank? || provider_id == "demo"

      result = provider_for(provider_id).verify(payment_id)
      status = result["status"]
      unless %w[paid canceled].include?(status)
        raise Infrastructure::JobDeferred, 60 if correlation_id.present?
        return
      end

      entity ||= recover_entity(provider_id, result["metadata"] || {})
      raise Infrastructure::JobDeferred, 60 if entity.nil? && correlation_id.present?
      raise Billing::BillingError, "Локальная привязка платежа ещё не готова." unless entity

      order_entity = entity.is_a?(Order)
      bound_by_id = entity.provider_payment_id == result["payment_id"]
      expected = result.dig("metadata", order_entity ? "order_id" : "topup_id")
      if !bound_by_id && provider_id == "platega" && expected != entity.id
        raise Billing::BillingError, "Платёж относится к другому заказу."
      end

      if order_entity && entity.provider_account.present? &&
         entity.provider_account != Infrastructure::RuntimeConfig.fetch("PLATEGA_MERCHANT_ID", "")
        raise Billing::BillingError, "Несовпадение магазина."
      end

      actual_id = result["payment_id"].to_s
      if actual_id.blank? || actual_id.length > 100 || (entity.provider_payment_id.present? && entity.provider_payment_id != actual_id)
        raise Billing::BillingError, "Несовпадение платежа."
      end

      if status == "paid"
        raise Billing::BillingError, "Тестовый платёж запрещён в production." if production? && result["test"]
        if order_entity
          @settlement.settle_order(entity.id, provider_id, actual_id, result["amount_kopeks"], result["currency"], correlation_id: correlation_id)
        else
          @topups.settle(entity.id, provider_id, actual_id, result["amount_kopeks"], result["currency"])
        end
        @attempts.completed(provider_id, actual_id, "paid")
      else
        entity.class.where(id: entity.id, provider: provider_id, status: "pending").update_all(status: "canceled")
        @attempts.completed(provider_id, actual_id, "canceled")
      end
    end

    def handle_webhook(provider_id, request)
      provider = provider_for(provider_id)
      return false unless provider.configured?

      result = provider.handle_webhook(request)
      return false unless result

      payment_id = result["payment_id"].to_s
      return false if payment_id.blank? || payment_id.length > 100

      status = result["status"]
      return true unless %w[paid canceled].include?(status)

      ApplicationRecord.transaction do
        event_id = result["event_id"].presence || "#{payment_id}:#{status}"
        # The webhook guard schema has a 64-character primary key. Keep the
        # provider event id stable while accepting valid long transaction IDs.
        event_id = Digest::SHA256.hexdigest(event_id) if event_id.length > 64
        begin
          claim = @webhook_guard.claim(provider_id, event_id, result)
        rescue WebhookTamperError => error
          Rails.logger.warn("webhook.tamper #{provider_id}/#{event_id} #{error.message}")
          next
        end
        next if claim == "duplicate"

        id = @events.receive(provider_id, event_id, payment_id, result, signature_valid: true, headers: request.headers.to_h)
        already = PaymentEvent.find(id)
        next if already.status == "processed"

        @outbox.enqueue("payment.event.process", "payment-event:#{id}", { event_id: id }, correlation_id: id)
      end
      true
    rescue Billing::BillingError
      false
    end

    def process_event(event_id)
      event = @events.claim(event_id)
      return unless event

      verify(event.payment_id, event.provider, correlation_id: event.id)
      @events.processed(event.id, event.lock_token)
      @webhook_guard.processed(event.provider, event.provider_event_id)
    rescue Infrastructure::JobDeferred => e
      @events.failed(event.id, event.lock_token, e) if event
      raise
    rescue Billing::BillingError
      if event
        @events.processed(event.id, event.lock_token)
        @webhook_guard.processed(event.provider, event.provider_event_id)
      end
    rescue StandardError => e
      @events.failed(event.id, event.lock_token, e) if event
      raise
    end

    # Reconcile external payments that cannot rely on a provider webhook and
    # replay webhook events whose outbox delivery or lease expired.
    def reconcile_pending(limit: 100, now: Time.now.to_i)
      provider = PlategaProvider.new
      if provider.configured?
        bucket = now / 300
        enqueue_pending_verifications(Order, "reconcile", bucket, limit)
        enqueue_pending_verifications(Topup, "reconcile-topup", bucket, limit)
      end

      stale_total = 0
      after = ""
      loop do
        stale_events = PaymentEvent.where(signature_valid: 1).where(
          "((status IN ('pending','retry') AND COALESCE(next_attempt_at, received_at) <= ?) OR (status = 'processing' AND locked_until <= ?)) AND id > ?",
          now, now, after
        ).order(:id).limit(limit)
        break if stale_events.empty?

        stale_events.each do |event|
          @outbox.enqueue("payment.event.process", "payment-event-recover:#{event.id}:#{now / 300}",
            { event_id: event.id }, correlation_id: event.id)
        end
        stale_total += stale_events.length
        break if stale_events.length < limit

        after = stale_events.last.id
      end
      stale_total
    end

    private

    def enqueue_pending_verifications(model, prefix, bucket, batch_size)
      relation = model.where(provider: "platega", status: "pending").where.not(provider_payment_id: [nil, ""])
      after = ""
      loop do
        rows = relation.where("id > ?", after).order(:id).limit(batch_size).pluck(:id, :provider_payment_id)
        break if rows.empty?

        rows.each do |id, payment_id|
          @outbox.enqueue("payment.verify", "#{prefix}:#{id}:#{bucket}",
            { payment_id: payment_id, provider: "platega" })
        end
        break if rows.length < batch_size

        after = rows.last.first
      end
    end

    def production?
      Rails.env.production? || Infrastructure::RuntimeConfig.fetch("APP_ENV", "") == "prod"
    end

    def provider_for(provider_id)
      return PlategaProvider.new if provider_id == "platega"

      raise Billing::BillingError, "Провайдер не поддерживается."
    end

    def scoped_payment_lookup(model, payment_id, provider_id)
      scope = model.where("provider_payment_id = ? OR id = ?", payment_id, payment_id)
      provider_id.present? ? scope.where(provider: provider_id).to_a : scope.to_a
    end

    def recover_entity(provider_id, metadata)
      order = Order.find_by(id: metadata["order_id"], provider: provider_id)
      topup = Topup.find_by(id: metadata["topup_id"], provider: provider_id)
      raise Billing::BillingError, "Неоднозначная привязка платежа." if order && topup

      order || topup
    end

    def existing_attempt_result(entity, attempt)
      if attempt.provider_payment_id.present? && attempt.checkout_url.present?
        entity.update!(provider_payment_id: attempt.provider_payment_id, checkout_url: attempt.checkout_url) if entity.checkout_url.blank?
        return { "payment_id" => attempt.provider_payment_id, "checkout_url" => attempt.checkout_url }
      end

      raise Billing::BillingError, "Платёжный checkout уже создаётся. Повторите позже."
    end

    def attach_order_checkout(order, attempt, payment_id, checkout_url)
      order.update!(provider_payment_id: payment_id, checkout_url: checkout_url)
      @attempts.attached(attempt.id, payment_id, checkout_url)
      { "payment_id" => payment_id, "checkout_url" => checkout_url }
    end

    def attach_topup_checkout(topup, attempt, payment_id, checkout_url)
      topup.update!(provider_payment_id: payment_id, checkout_url: checkout_url)
      @attempts.attached(attempt.id, payment_id, checkout_url)
      { "payment_id" => payment_id, "checkout_url" => checkout_url }
    end
  end
end
